### ISSUE TITLE
Agent and Working State

### ISSUE THREAD

**Issue by ssavva05** [hpcn-uam/q4s#2]
Hello,
 I try to find a q4s implementation that is completed. Is there any stable version of Q4s that i can  use it as a protocol to make my own decision machine. Is that implementation be an emulator? Can this emulator work as expected?

**Comment by dmuelasr** [comment]
Hello,

This is a complete implementation of the Q4S protocol (server and client) with some additional customizable parameters. Currently, we are concluding the validation and integration of this software with other third-party elements, so we plan to wait until this process is finished before labeling a stable version.

Anyway, our tests so far showed that it is fully operational and can be executed in GNU/Linux systems with modest hardware. Hence, I believe that you could use the software to deploy it in your own testbed. Note that the Agent is a legacy and discontinued element: you need only the client and server parts, and you could also adapt the sample actuator.

**Comment by ssavva05** [comment]
Great! Thanks for that valuable information! I try to bottleneck the network using two Debian Virtual Machines and using tc command and the printed results shows only that we have packet losseses. (I add jitter and bandwith limits and I get only packet losses):
 
![image](https://user-images.githubusercontent.com/22216008/47173212-26e3ba80-d316-11e8-8013-255e295f6822.png)

( I also want to know about the server in the port 8080 if it is possible. When I run the implementation I get error about curl in an ip and port 8080).

Is there any documentation about this implementation?

**Comment by dmuelasr** [comment]
Regarding the measurements, I would need some more info about the configurations of both tc and Q4S: if the BW limitation is much lower than the BW set in the configuration of Q4S server, you may see an abrupt level of packet loss with bad estimations of the rest of parameters.

Regarding the server in port 8080, we have a web-based actuator (in the repo), which is application-specific (and out of the scope of Q4S). My guess is that the Q4S element is sending an alert to the actuator, which in your case is not running. This part of the project is quite an ad hoc solution yet, but as said before we are finishing the integration with third party elements, so we expect to polish it soon :)

Finally, documentation is in a premature state. We have tried to add some help to compile and run the solution (readme and wiki), but we are using this software in a a project and we need to finish some other parts before having the time to prepare more detailed documentations. Anyway, we will try to upload some further details in the next couple of months to close this part.

**Comment by ssavva05** [comment]
Great!  For the Virtual Machine Settings . I have two virtual machines running Debian: One Client and One Server. 
 I use sudo tc qdisc add dev eth0 root tbf rate 64kbit latency 250ms burst 1540 only in client machine. 
To be more specific , I have to use Q4S for a semester project in computer networking class. 
 I want to implement on top of it a decision machine and a streaming server (probably RTP).
![image](https://user-images.githubusercontent.com/22216008/47178154-04a46980-d323-11e8-9e0b-7ae567eb8e39.png)
I have to find the modules and put it together. 

I try to test also this implementation https://github.com/gmunozfe/Q4S_Stack_JS_Chrome_Extension ,but i am stuck. I don't know how to run its server...

I hope that i can use this implementation in Q4S part.... I want to hear your opinion is it good to continue with this implementation?

**Comment by dmuelasr** [comment]
Well, from my standpoint if you do not have to implement the protocol for your assignment, our version is quite stable and we have also assessed with tc/netem how it estimates the network QoS. You only need to configure the Q4S server (see instructions in the readme or wiki) to reflect the actual requirements of the application, and the decision machine should be included in the actuator module (we have a sample web-based one, which we used to configure the application and include the specific policies of a decision module).

Importantly, if network QoS (i.e., BW, latency, jitter, and loss limitations included with tc/netem) are far below the requirements of the application, Q4S will send alerts to the actuator and change the quality level until fullfiling the constraints.  If no quality level of the application suits the network conditions, the application may not exectute. That means that in your scenario, the streaming application must offer a configuration that can execute with the network characteristics included in netem. Then, that configuration has to be linked to a quality level for Q4S, which will notify when the corresponding constraints are met.

I hope this can help!
